crates/client/src/lib.rs

```diff
@@ -574,71 +557,169 @@ impl<R: RegistryStorage, C: ContentStorage, N: NamespaceMapStorage> Client<R, C,
         Ok(PackageDownload {
             version: version.clone(),
             digest: digest.clone(),
-            path: self.download_content(digest).await?,
+            path: self
+                .download_content(registry_domain.as_ref(), digest)
+                .await?,
         })
     }
 
-    /// Update checkpoint for list of packages
-    async fn update_checkpoint<'a>(
+    async fn update_packages_and_return_federated_packages<'a>(
         &self,
-        ts_checkpoint: &SerdeEnvelope<TimestampedCheckpoint>,
-        packages: impl IntoIterator<Item = &mut PackageInfo>,
-    ) -> Result<(), ClientError> {
+        registry_domain: Option<&RegistryDomain>,
+        packages: impl IntoIterator<Item = &'a mut PackageInfo>,
+    ) -> Result<IndexMap<Option<RegistryDomain>, Vec<&'a mut PackageInfo>>, ClientError> {
+        let ts_checkpoint = self.api.latest_checkpoint(registry_domain).await?;
         let checkpoint = &ts_checkpoint.as_ref().checkpoint;
-        tracing::info!(
-            "updating to checkpoint log length `{}`",
-            checkpoint.log_length
-        );
 
+        if let Some(registry_header) = registry_domain {
+            tracing::info!(
+                "updating to checkpoint log length `{}` with registry header: {registry_header}",
+                checkpoint.log_length
+            );
+        } else {
+            tracing::info!(
+                "updating to checkpoint log length `{}`",
+                checkpoint.log_length
+            );
+        }
+
+        // operator log info
         let mut operator = self
             .registry
-            .load_operator(self.api.get_warg_registry())
+            .load_operator(registry_domain)
             .await?
             .unwrap_or_default();
 
-        // Map package names to package logs that need to be updated
+        // map package names to package logs that need to be updated
         let mut packages = packages
             .into_iter()
             .filter_map(|p| match &p.checkpoint {
                 // Don't bother updating if the package is already at the specified checkpoint
-                Some(c) if c == checkpoint => None,
+                // If `registry` field is not set, then update.
+                Some(c) if p.registry.is_some() && c == checkpoint => None,
```

Comment:
Is it not possible for this function to be called with an out of date package that has not had a registry set?  Feels this could cause trouble if somebody doesn't remember to set the registry on a package

Reply:
Discussed over a call.